File: go_src/internal/hillclimb/hillclimb.go
// Package hillclimb implements the hill-climb-like local-search SAT
// solving methods used by vibe_sat:
//
//   - Run implements the simple hill-climb from STAGE2.md: repeatedly
//     start from a random complete assignment and greedily flip single
//     variables as long as doing so increases the number of satisfied
//     clauses, restarting whenever no single flip can improve further.
//   - RunWalkSat implements WalkSAT (STAGE4.md's more advanced
//     hill-climb), which always flips a variable from some currently
//     unsatisfied clause -- usually the one that breaks the fewest
//     other clauses, but occasionally a random one -- which lets it
//     escape the local optima that can trap the simple hill-climb.
//
// Both algorithms are built on the same climbState, which tracks a
// complete assignment together with enough per-clause bookkeeping to
// evaluate and apply a single-variable flip in time proportional to
// that variable's occurrence count.
package hillclimb

import (
	"fmt"
	"math/rand/v2"
	"sync"
	"sync/atomic"
	"time"

	"vibe_sat/internal/assign"
	"vibe_sat/internal/cnf"
	"vibe_sat/internal/occurrence"
)

// Params configures a single call to Run: how many random restarts to
// attempt, how long to keep searching, or both (in which case the
// search stops as soon as either limit is reached). A nil field means
// that limit does not apply.
//
// Stop and BestScore (STAGE17.md) exist to let RunParallel coordinate
// several concurrently running instances of the same search; every
// single-threaded caller (including Run itself) leaves them nil, in
// which case they have no effect whatsoever and behavior is identical
// to before Stage 17.
type Params struct {
	NumStarts *int
	TimeLimit *time.Duration

	// Stop, if non-nil, is checked before every restart; if it is
	// already true, the search stops immediately, exactly as if
	// NumStarts/TimeLimit had been reached. RunParallel gives every
	// worker the same *atomic.Bool, so that the moment any one of them
	// finds a satisfying assignment, every other worker notices at its
	// next restart and stops promptly instead of continuing to search
	// for a solution that is no longer needed.
	Stop *atomic.Bool

	// BestScore, if non-nil, replaces Run's own local best-score
	// tracking: instead of comparing against a private variable,
	// recordIfBest compare-and-swaps against this shared value, so
	// that several concurrently running instances (RunParallel's
	// workers) track one global best score between them, and each
	// only prints "new best score" (verbose >= 2) when it actually
	// raised that global maximum -- not merely its own, possibly
	// stale, local one.
	BestScore *atomic.Int64
}

// Result describes the outcome of a hill-climbing run.
type Result struct {
	Satisfiable bool              // whether a fully satisfying assignment was found
	Assignment  assign.Assignment // the satisfying assignment; only meaningful if Satisfiable
	Score       int               // the score of Assignment (number of satisfied clauses)
	Starts      int               // number of random restarts performed
}

// Score returns the number of clauses in problem that are satisfied
// by assignment.
func Score(problem *cnf.Problem, assignment assign.Assignment) int {
	score := 0
	for _, clause := range problem.Clauses {
		if clauseIsSatisfied(clause, assignment) {
			score++
		}
	}
	return score
}

// clauseIsSatisfied reports whether at least one literal of clause is
// true under assignment.
func clauseIsSatisfied(clause cnf.Clause, assignment assign.Assignment) bool {
	for _, lit := range clause {
		if assignment.LiteralIsTrue(lit) {
			return true
		}
	}
	return false
}

// clauseTrueCounts computes, for every clause of problem, the number
// of its literals that are currently true under assignment, along
// with the resulting score (the number of clauses whose count is
// greater than zero, i.e. satisfied).
func clauseTrueCounts(problem *cnf.Problem, assignment assign.Assignment) (counts []int, score int) {
	counts = make([]int, len(problem.Clauses))
	for i, clause := range problem.Clauses {
		for _, lit := range clause {
			if assignment.LiteralIsTrue(lit) {
				counts[i]++
			}
		}
		if counts[i] > 0 {
			score++
		}
	}
	return counts, score
}

// climbState holds the mutable bookkeeping needed to evaluate and
// apply single-variable flips in time proportional to that variable's
// occurrence count, rather than rescanning every clause on every
// flip. It is shared by every local-search algorithm in this package
// (the simple hill-climb in Run, and WalkSAT in RunWalkSat), since
// both are built on the same incremental satisfied-clause tracking.
type climbState struct {
	problem      *cnf.Problem
	lists        *occurrence.Lists
	assignment   assign.Assignment
	counts       []int // counts[c] = number of true literals in clause c
	score        int   // number of clauses with counts[c] > 0
	unsatClauses []int // indices of every clause currently unsatisfied, in no particular order
	unsatPos     []int // unsatPos[c] = index of c within unsatClauses, or -1 if c is satisfied
}

// newClimbState builds a climbState from a complete starting
// assignment, computing the initial per-clause true-literal counts,
// score, and unsatisfied-clause set from scratch.
func newClimbState(problem *cnf.Problem, lists *occurrence.Lists, assignment assign.Assignment) *climbState {
	counts, score := clauseTrueCounts(problem, assignment)
	s := &climbState{
		problem:    problem,
		lists:      lists,
		assignment: assignment,
		counts:     counts,
		score:      score,
		unsatPos:   make([]int, len(problem.Clauses)),
	}
	for c, count := range counts {
		if count == 0 {
			s.unsatPos[c] = len(s.unsatClauses)
			s.unsatClauses = append(s.unsatClauses, c)
		} else {
			s.unsatPos[c] = -1
		}
	}
	return s
}

// markUnsatisfied records that clause c has just become unsatisfied,
// in O(1).
func (s *climbState) markUnsatisfied(c int) {
	s.unsatPos[c] = len(s.unsatClauses)
	s.unsatClauses = append(s.unsatClauses, c)
}

// markSatisfied records that clause c has just become satisfied, in
// O(1), by swapping it with the last entry of unsatClauses before
// shrinking the slice.
func (s *climbState) markSatisfied(c int) {
	pos := s.unsatPos[c]
	last := len(s.unsatClauses) - 1
	movedClause := s.unsatClauses[last]
	s.unsatClauses[pos] = movedClause
	s.unsatPos[movedClause] = pos
	s.unsatClauses = s.unsatClauses[:last]
	s.unsatPos[c] = -1
}

// flip flips the value of variable v in place, updates the per-clause
// true-literal counts, the overall score, and the unsatisfied-clause
// set to match, and returns the resulting change in score. Calling
// flip a second time with the same v exactly reverses the first call,
// since flipping is its own inverse.
func (s *climbState) flip(v int) int {
	wasTrue := s.assignment[v] == assign.True

	var losing, gaining []int
	if wasTrue {
		losing, gaining = s.lists.Positive[v], s.lists.Negative[v]
	} else {
		losing, gaining = s.lists.Negative[v], s.lists.Positive[v]
	}

	delta := 0
	for _, c := range losing {
		s.counts[c]--
		if s.counts[c] == 0 {
			delta--
			s.markUnsatisfied(c)
		}
	}
	for _, c := range gaining {
		s.counts[c]++
		if s.counts[c] == 1 {
			delta++
			s.markSatisfied(c)
		}
	}

	if wasTrue {
		s.assignment[v] = assign.False
	} else {
		s.assignment[v] = assign.True
	}
	s.score += delta
	return delta
}

// breakCount returns the number of currently satisfied clauses that
// would become unsatisfied if variable v were flipped right now,
// without actually flipping it. WalkSAT (see walksat.go) uses this to
// prefer flips that break as few other clauses as possible.
func (s *climbState) breakCount(v int) int {
	wasTrue := s.assignment[v] == assign.True

	var losing []int
	if wasTrue {
		losing = s.lists.Positive[v]
	} else {
		losing = s.lists.Negative[v]
	}

	count := 0
	for _, c := range losing {
		if s.counts[c] == 1 {
			count++
		}
	}
	return count
}

// sweep performs one pass over the variables in the order given by
// order, attempting to flip each one in turn. A flip that strictly
// increases the score is kept, in which case onImproved (if non-nil)
// is invoked with the new score; a flip that does not strictly
// improve the score is reverted before moving on. sweep returns true
// if at least one flip was kept during the pass.
func (s *climbState) sweep(order []int, onImproved func(newScore int)) bool {
	changed := false
	for _, v := range order {
		delta := s.flip(v)
		if delta > 0 {
			changed = true
			if onImproved != nil {
				onImproved(s.score)
			}
		} else {
			s.flip(v) // revert
		}
	}
	return changed
}

// randomPermutation returns a random permutation of the variable
// numbers 1 through numVars, ordered using rng.
func randomPermutation(numVars int, rng *rand.Rand) []int {
	order := make([]int, numVars)
	for i := range order {
		order[i] = i + 1
	}
	rng.Shuffle(numVars, func(i, j int) { order[i], order[j] = order[j], order[i] })
	return order
}

// Run performs the hill-climbing search described by STAGE2.md:
// repeatedly starting from a random complete assignment and greedily
// flipping variables until no single flip can improve the score,
// stopping as soon as a satisfying assignment is found or the limits
// in params are reached. rng supplies all the randomness used (the
// initial assignment and flip order of every start), and verbose
// controls how much progress is printed to stdout:
//
//   - verbose >= 1: prints "hillclimb" with the configured limits
//     before searching, and "SAT" or "UNKNOWN" after searching.
//   - verbose >= 2: prints the score every time it sets a new best
//     score across all starts so far.
//   - verbose >= 3: prints the score every time a start gets stuck at
//     a local optimum that does not satisfy every clause.
func Run(problem *cnf.Problem, lists *occurrence.Lists, params Params, rng *rand.Rand, verbose int) Result {
	if verbose >= 1 {
		fmt.Println("hillclimb:", describeParams(params))
	}
	result := runLoop(problem, lists, params, rng, verbose)
	if verbose >= 1 {
		fmt.Println(verdict(result.Satisfiable))
	}
	return result
}

// RunParallel runs the same search as Run, split across numThreads
// concurrent workers (STAGE17.md). numThreads <= 1 delegates straight
// to Run, with rng used exactly as it always has been -- so behavior
// (including every random choice made) is bit-for-bit identical to
// calling Run directly whenever multithreading isn't actually in use.
//
// If params.NumStarts is given, it is split as evenly as possible
// across the workers (each doing ceil(NumStarts/numThreads) starts,
// so the total may run a few more starts than asked for, never
// fewer); params.TimeLimit, if given, is handed to every worker
// unchanged rather than divided, since the workers run concurrently
// and dividing it would just shorten the wall-clock time spent
// without letting any more work fit in it. rng is used, before any
// worker starts, to derive one independent sub-generator per worker
// (see newSubRand), so the overall result is fully reproducible given
// (rng's state, numThreads) even though which worker's answer wins a
// race to a solution is not.
//
// The moment any worker finds a satisfying assignment, it signals
// every other worker to stop at its next restart (see Params.Stop),
// and that worker's result -- and only that worker's -- is the one
// RunParallel reports; every other worker's own result, even a
// simultaneously-successful one, is discarded, so exactly one
// solution is ever reported, per STAGE17.md. Result.Starts is the sum
// of every worker's own start count, win or lose.
func RunParallel(problem *cnf.Problem, lists *occurrence.Lists, params Params, numThreads int, rng *rand.Rand, verbose int) Result {
	if numThreads <= 1 {
		return Run(problem, lists, params, rng, verbose)
	}

	if verbose >= 1 {
		fmt.Println("hillclimb:", describeParallelParams(params, numThreads))
	}

	subRands := make([]*rand.Rand, numThreads)
	for i := range subRands {
		subRands[i] = newSubRand(rng)
	}

	var stop atomic.Bool
	var bestScore atomic.Int64
	bestScore.Store(-1)
	var winner atomic.Int32
	winner.Store(-1)

	results := make([]Result, numThreads)
	var wg sync.WaitGroup
	for i := 0; i < numThreads; i++ {
		workerParams := params
		workerParams.Stop = &stop
		workerParams.BestScore = &bestScore
		if params.NumStarts != nil {
			n := ceilDiv(*params.NumStarts, numThreads)
			workerParams.NumStarts = &n
		}

		wg.Add(1)
		go func(i int) {
			defer wg.Done()
			results[i] = runLoop(problem, lists, workerParams, subRands[i], verbose)
			if results[i].Satisfiable && stop.CompareAndSwap(false, true) {
				winner.Store(int32(i))
			}
		}(i)
	}
	wg.Wait()

	final := Result{}
	for _, r := range results {
		final.Starts += r.Starts
	}
	if idx := winner.Load(); idx >= 0 {
		final.Satisfiable = true
		final.Assignment = results[idx].Assignment
		final.Score = results[idx].Score
	}

	if verbose >= 1 {
		fmt.Println(verdict(final.Satisfiable))
	}
	return final
}

// runLoop is Run's search loop, without the "hillclimb: ..."/
// "SAT"/"UNKNOWN" announcements -- those are the caller's
// responsibility (Run prints its own; RunParallel prints one combined
// announcement/verdict for the whole parallel search instead of one
// per worker), so that multithreaded runs don't produce numThreads
// duplicate announcement lines.
func runLoop(problem *cnf.Problem, lists *occurrence.Lists, params Params, rng *rand.Rand, verbose int) Result {
	startTime := time.Now()
	bestScore := -1
	recordIfBest := func(score int) {
		if params.BestScore != nil {
			for {
				old := params.BestScore.Load()
				if int64(score) <= old {
					return
				}
				if params.BestScore.CompareAndSwap(old, int64(score)) {
					if verbose >= 2 {
						fmt.Printf("new best score: %d/%d\n", score, problem.NumClauses())
					}
					return
				}
				// Another worker updated it concurrently; retry against
				// whatever the new value is.
			}
		}
		if score > bestScore {
			bestScore = score
			if verbose >= 2 {
				fmt.Printf("new best score: %d/%d\n", score, problem.NumClauses())
			}
		}
	}

	starts := 0
	for {
		if params.Stop != nil && params.Stop.Load() {
			break
		}
		if params.NumStarts != nil && starts >= *params.NumStarts {
			break
		}
		if params.TimeLimit != nil && time.Since(startTime) >= *params.TimeLimit {
			break
		}
		starts++

		assignment := assign.NewRandom(problem.NumVars, rng)
		state := newClimbState(problem, lists, assignment)
		recordIfBest(state.score)

		for {
			order := randomPermutation(problem.NumVars, rng)
			if !state.sweep(order, recordIfBest) {
				break
			}
		}

		if state.score == problem.NumClauses() {
			return Result{Satisfiable: true, Assignment: state.assignment, Score: state.score, Starts: starts}
		}

		if verbose >= 3 {
			fmt.Printf("stuck at score %d/%d after start %d\n", state.score, problem.NumClauses(), starts)
		}
	}

	return Result{Satisfiable: false, Starts: starts}
}

// newSubRand draws two fresh uint64s from rng to seed a new,
// independent *rand.Rand. Used by RunParallel to give every worker
// its own generator (drained sequentially from the caller's rng,
// before any worker goroutine starts, so the whole sequence of
// derived seeds is itself deterministic given rng's own state) rather
// than sharing one *rand.Rand across goroutines, which would be both
// a data race and a source of scheduling-dependent nondeterminism.
func newSubRand(rng *rand.Rand) *rand.Rand {
	return rand.New(rand.NewPCG(rng.Uint64(), rng.Uint64()))
}

// ceilDiv returns ceil(a/b) for positive a and b, without floating
// point.
func ceilDiv(a, b int) int {
	return (a + b - 1) / b
}

// verdict returns "SAT" or "UNKNOWN", matching what Run/RunWalkSat
// print at verbose level 1 once a search concludes.
func verdict(satisfiable bool) string {
	if satisfiable {
		return "SAT"
	}
	return "UNKNOWN"
}

// describeParams formats the configured restart/time limits for the
// "hillclimb:" announcement printed at verbose level 1.
func describeParams(params Params) string {
	description := ""
	if params.NumStarts != nil {
		description += fmt.Sprintf("num_starts=%d ", *params.NumStarts)
	}
	if params.TimeLimit != nil {
		description += fmt.Sprintf("time_limit_secs=%d ", int(params.TimeLimit.Seconds()))
	}
	if description == "" {
		return "(no limit)"
	}
	return description[:len(description)-1]
}

// describeParallelParams is describeParams, extended with the worker
// count, for RunParallel's "hillclimb: ..." announcement.
func describeParallelParams(params Params, numThreads int) string {
	return fmt.Sprintf("num_threads=%d %s", numThreads, describeParams(params))
}

File: rust_src/src/hillclimb.rs
//! Implements the simplest local-search SAT solving method used by
//! vibe_sat: repeatedly start from a random complete assignment and
//! greedily flip single variables as long as doing so increases the
//! number of satisfied clauses, restarting from a new random
//! assignment whenever no single flip can improve further.

use std::time::{Duration, Instant};

use rand::Rng;
use rand::seq::SliceRandom;

use crate::assignment::{self, Assignment, Value};
use crate::cnf::{Clause, Problem};
use crate::occurrence::Lists;

/// Configures a single call to [`run`]: how many random restarts to
/// attempt, how long to keep searching, or both (in which case the
/// search stops as soon as either limit is reached). `None` means
/// that limit does not apply.
#[derive(Debug, Clone, Copy, Default)]
pub struct Params {
    pub num_starts: Option<usize>,
    pub time_limit: Option<Duration>,
}

/// Describes the outcome of a hill-climbing run.
pub struct SolveResult {
    /// Whether a fully satisfying assignment was found.
    pub satisfiable: bool,
    /// The satisfying assignment; only meaningful if `satisfiable`.
    pub assignment: Assignment,
    /// The score of `assignment` (number of satisfied clauses). Not
    /// read by `main` in this stage, but kept as diagnostic output
    /// for callers (and covered directly by tests).
    #[allow(dead_code)]
    pub score: usize,
    /// The number of random restarts performed. Same status as
    /// `score` above.
    #[allow(dead_code)]
    pub starts: usize,
}

/// Returns whether at least one literal of `clause` is true under
/// `assignment`.
fn clause_is_satisfied(clause: &Clause, assignment: &Assignment) -> bool {
    clause
        .iter()
        .any(|&literal| assignment::literal_is_true(assignment, literal))
}

/// Returns the number of clauses in `problem` that are satisfied by
/// `assignment`. Not called outside of tests in this stage (the
/// hill-climb loop tracks its score incrementally instead), but kept
/// as a straightforward, independently-testable reference
/// implementation that the incremental bookkeeping is checked
/// against.
#[allow(dead_code)]
pub fn score(problem: &Problem, assignment: &Assignment) -> usize {
    problem
        .clauses
        .iter()
        .filter(|clause| clause_is_satisfied(clause, assignment))
        .count()
}

/// Computes, for every clause of `problem`, the number of its
/// literals that are currently true under `assignment`, along with
/// the resulting score (the number of clauses whose count is greater
/// than zero, i.e. satisfied).
fn clause_true_counts(problem: &Problem, assignment: &Assignment) -> (Vec<usize>, usize) {
    let mut counts = vec![0usize; problem.clauses.len()];
    let mut current_score = 0;
    for (i, clause) in problem.clauses.iter().enumerate() {
        for &literal in clause {
            if assignment::literal_is_true(assignment, literal) {
                counts[i] += 1;
            }
        }
        if counts[i] > 0 {
            current_score += 1;
        }
    }
    (counts, current_score)
}

/// Holds the mutable bookkeeping needed to evaluate and apply
/// single-variable flips in time proportional to that variable's
/// occurrence count, rather than rescanning every clause on every
/// flip.
struct ClimbState<'a> {
    lists: &'a Lists,
    assignment: Assignment,
    counts: Vec<usize>, // counts[c] = number of true literals in clause c
    score: usize,       // number of clauses with counts[c] > 0
}

impl<'a> ClimbState<'a> {
    /// Builds a `ClimbState` from a complete starting assignment,
    /// computing the initial per-clause true-literal counts and score
    /// from scratch.
    fn new(problem: &Problem, lists: &'a Lists, assignment: Assignment) -> Self {
        let (counts, score) = clause_true_counts(problem, &assignment);
        ClimbState {
            lists,
            assignment,
            counts,
            score,
        }
    }

    /// Flips the value of variable `v` in place, updates the
    /// per-clause true-literal counts and the overall score to match,
    /// and returns the resulting change in score. Calling `flip` a
    /// second time with the same `v` exactly reverses the first call,
    /// since flipping is its own inverse.
    fn flip(&mut self, v: usize) -> i64 {
        let was_true = self.assignment[v] == Value::True;
        let (losing, gaining) = if was_true {
            (&self.lists.positive[v], &self.lists.negative[v])
        } else {
            (&self.lists.negative[v], &self.lists.positive[v])
        };

        let mut delta: i64 = 0;
        for &c in losing {
            self.counts[c] -= 1;
            if self.counts[c] == 0 {
                delta -= 1;
            }
        }
        for &c in gaining {
            self.counts[c] += 1;
            if self.counts[c] == 1 {
                delta += 1;
            }
        }

        self.assignment[v] = if was_true { Value::False } else { Value::True };
        self.score = (self.score as i64 + delta) as usize;
        delta
    }

    /// Performs one pass over the variables in the order given by
    /// `order`, attempting to flip each one in turn. A flip that
    /// strictly increases the score is kept, in which case
    /// `on_improved` is invoked with the new score; a flip that does
    /// not strictly improve the score is reverted before moving on.
    /// Returns true if at least one flip was kept during the pass.
    fn sweep(&mut self, order: &[usize], mut on_improved: impl FnMut(usize)) -> bool {
        let mut changed = false;
        for &v in order {
            let delta = self.flip(v);
            if delta > 0 {
                changed = true;
                on_improved(self.score);
            } else {
                self.flip(v); // revert
            }
        }
        changed
    }
}

/// Returns a random permutation of the variable numbers 1 through
/// `num_vars`, ordered using `rng`.
fn random_permutation<R: Rng>(num_vars: usize, rng: &mut R) -> Vec<usize> {
    let mut order: Vec<usize> = (1..=num_vars).collect();
    order.shuffle(rng);
    order
}

/// Formats the configured restart/time limits for the "hillclimb:"
/// announcement printed at verbose level 1.
fn describe_params(params: Params) -> String {
    let mut parts = Vec::new();
    if let Some(n) = params.num_starts {
        parts.push(format!("num_starts={n}"));
    }
    if let Some(limit) = params.time_limit {
        parts.push(format!("time_limit_secs={}", limit.as_secs()));
    }
    if parts.is_empty() {
        "(no limit)".to_string()
    } else {
        parts.join(" ")
    }
}

/// Performs the hill-climbing search described by STAGE2.md:
/// repeatedly starting from a random complete assignment and greedily
/// flipping variables until no single flip can improve the score,
/// stopping as soon as a satisfying assignment is found or the limits
/// in `params` are reached. `rng` supplies all the randomness used
/// (the initial assignment and flip order of every start), and
/// `verbose` controls how much progress is printed to stdout:
///
/// - `verbose >= 1`: prints "hillclimb" with the configured limits
///   before searching, and "SAT" or "UNKNOWN" after searching.
/// - `verbose >= 2`: prints the score every time it sets a new best
///   score across all starts so far.
/// - `verbose >= 3`: prints the score every time a start gets stuck
///   at a local optimum that does not satisfy every clause.
pub fn run<R: Rng>(
    problem: &Problem,
    lists: &Lists,
    params: Params,
    rng: &mut R,
    verbose: i32,
) -> SolveResult {
    if verbose >= 1 {
        println!("hillclimb: {}", describe_params(params));
    }

    let start_time = Instant::now();
    let mut best_score: i64 = -1;
    let num_clauses = problem.num_clauses();

    let mut record_if_best = |score: usize| {
        if score as i64 > best_score {
            best_score = score as i64;
            if verbose >= 2 {
                println!("new best score: {score}/{num_clauses}");
            }
        }
    };

    let mut starts = 0usize;
    loop {
        if let Some(limit) = params.num_starts
            && starts >= limit
        {
            break;
        }
        if let Some(limit) = params.time_limit
            && start_time.elapsed() >= limit
        {
            break;
        }
        starts += 1;

        let initial_assignment = assignment::new_random(problem.num_vars, rng);
        let mut state = ClimbState::new(problem, lists, initial_assignment);
        record_if_best(state.score);

        loop {
            let order = random_permutation(problem.num_vars, rng);
            if !state.sweep(&order, &mut record_if_best) {
                break;
            }
        }

        if state.score == num_clauses {
            if verbose >= 1 {
                println!("SAT");
            }
            return SolveResult {
                satisfiable: true,
                assignment: state.assignment,
                score: state.score,
                starts,
            };
        }

        if verbose >= 3 {
            println!(
                "stuck at score {}/{num_clauses} after start {starts}",
                state.score
            );
        }
    }

    if verbose >= 1 {
        println!("UNKNOWN");
    }
    SolveResult {
        satisfiable: false,
        assignment: assignment::new(problem.num_vars),
        score: 0,
        starts,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::occurrence;
    use rand::SeedableRng;
    use rand::rngs::StdRng;

    #[test]
    fn test_score_counts_satisfied_clauses() {
        let problem = Problem {
            num_vars: 2,
            clauses: vec![vec![1], vec![-1], vec![2, -1]],
        };
        let mut a = assignment::new(2);
        a[1] = Value::True;
        a[2] = Value::False;

        // Clause {1} is satisfied (1 is True). Clause {-1} and clause
        // {2, -1} are both unsatisfied (1 is True, so -1 is false; 2
        // is False). Only 1 of the 3 clauses is satisfied.
        assert_eq!(score(&problem, &a), 1);
    }

    #[test]
    fn test_clause_true_counts_matches_score() {
        let problem = Problem {
            num_vars: 2,
            clauses: vec![vec![1, 2], vec![-1, -2]],
        };
        let mut a = assignment::new(2);
        a[1] = Value::True;
        a[2] = Value::True;

        let (counts, computed_score) = clause_true_counts(&problem, &a);
        assert_eq!(computed_score, score(&problem, &a));
        assert_eq!(counts[0], 2);
        assert_eq!(counts[1], 0);
    }

    #[test]
    fn test_flip_updates_score_correctly() {
        let problem = Problem {
            num_vars: 2,
            clauses: vec![vec![1, 2], vec![-1]],
        };
        let lists = occurrence::build(&problem);
        let mut a = assignment::new(2);
        a[1] = Value::False;
        a[2] = Value::False;

        let mut state = ClimbState::new(&problem, &lists, a);
        let delta = state.flip(1); // variable 1: False -> True

        let (want_counts, want_score) = clause_true_counts(&problem, &state.assignment);
        assert_eq!(state.score, want_score);
        assert_eq!(state.counts, want_counts);
        assert_eq!(state.assignment[1], Value::True);
        // Clause 0 goes from unsatisfied to satisfied (+1), clause 1
        // goes from satisfied to unsatisfied (-1): net delta 0.
        assert_eq!(delta, 0);
    }

    #[test]
    fn test_flip_twice_reverts_to_original_state() {
        let problem = Problem {
            num_vars: 3,
            clauses: vec![vec![1, -2], vec![2, 3], vec![-3]],
        };
        let lists = occurrence::build(&problem);
        let a = assignment::new_random(3, &mut StdRng::seed_from_u64(9));
        let original_assignment = a.clone();

        let mut state = ClimbState::new(&problem, &lists, a);
        let original_counts = state.counts.clone();
        let original_score = state.score;

        state.flip(2);
        state.flip(2);

        assert_eq!(state.assignment, original_assignment);
        assert_eq!(state.counts, original_counts);
        assert_eq!(state.score, original_score);
    }

    #[test]
    fn test_flip_handles_tautological_clause() {
        let problem = Problem {
            num_vars: 1,
            clauses: vec![vec![1, -1]],
        };
        let lists = occurrence::build(&problem);
        let mut a = assignment::new(1);
        a[1] = Value::True;

        let mut state = ClimbState::new(&problem, &lists, a);
        assert_eq!(state.score, 1);

        let delta = state.flip(1);
        assert_eq!(delta, 0);
        assert_eq!(state.score, 1);
        assert_eq!(state.counts[0], 1);
    }

    #[test]
    fn test_run_finds_satisfiable_formula() {
        // Every clause is a single positive literal, so flipping any
        // currently-false variable to true strictly helps; hill
        // climbing is guaranteed to reach full satisfaction.
        let problem = Problem {
            num_vars: 5,
            clauses: vec![vec![1], vec![2], vec![3], vec![4], vec![5]],
        };
        let lists = occurrence::build(&problem);
        let params = Params {
            num_starts: Some(1),
            time_limit: None,
        };
        let mut rng = StdRng::seed_from_u64(1);

        let result = run(&problem, &lists, params, &mut rng, 0);

        assert!(
            result.satisfiable,
            "score {}/{}",
            result.score,
            problem.num_clauses()
        );
        assert_eq!(result.score, problem.num_clauses());
    }

    #[test]
    fn test_run_reports_unknown_for_unsatisfiable_formula() {
        // x1 AND NOT x1: never satisfiable, regardless of restarts.
        let problem = Problem {
            num_vars: 1,
            clauses: vec![vec![1], vec![-1]],
        };
        let lists = occurrence::build(&problem);
        let params = Params {
            num_starts: Some(5),
            time_limit: None,
        };
        let mut rng = StdRng::seed_from_u64(2);

        let result = run(&problem, &lists, params, &mut rng, 0);

        assert!(!result.satisfiable);
        assert_eq!(result.starts, 5);
    }

    #[test]
    fn test_random_permutation_is_a_permutation() {
        let mut rng = StdRng::seed_from_u64(3);
        let order = random_permutation(10, &mut rng);
        let mut seen = std::collections::HashSet::new();
        for &v in &order {
            assert!((1..=10).contains(&v));
            assert!(seen.insert(v), "duplicate value {v}");
        }
        assert_eq!(seen.len(), 10);
    }
}
